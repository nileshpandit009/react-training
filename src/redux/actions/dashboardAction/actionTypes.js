export const DASHBOARD_CONSTANTS = {
  FETCH_POSTS: "FETCH_POSTS",
  FETCH_POSTS_SUCCESS: "FETCH_POSTS_SUCCESS",
  FETCH_POSTS_ERROR: "FETCH_POSTS_ERROR",

  FETCH_POST_DETAILS: "FETCH_POST_DETAILS",
  FETCH_POST_DETAILS_SUCCESS: "FETCH_POST_DETAILS_SUCCESS",
  FETCH_POST_DETAILS_ERROR: "FETCH_POST_DETAILS_ERROR",

  RESET_POSTS: "RESET_POSTS",
  RESET_POST_DETAILS: "RESET_POST_DETAILS",
};
