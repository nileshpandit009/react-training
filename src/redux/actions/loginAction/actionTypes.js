export const LOGIN_CONSTANTS = {
  EMAIL: "EMAIL",
  PASSWORD: "PASSWORD",
  EMAIL_ERR: "EMAIL_ERR",
  PASSWORD_ERR: "PASSWORD_ERR",
  LOGIN_ERR: "LOGIN_ERR",
  RESET_LOGIN: "RESET_LOGIN",
  LOGIN_REQUEST: "LOGIN_REQUEST",
  LOGIN_SUCCESS: "LOGIN_SUCCESS",
  LOGIN_FAILURE: "LOGIN_FAILURE",
};
