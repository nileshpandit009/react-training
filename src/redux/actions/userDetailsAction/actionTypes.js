export const USER_DETAILS_CONSTANTS = {
  USER_DETAILS_REQUEST: "USER_DETAILS_REQUEST",
  USER_DETAILS_SUCCESS: "USER_DETAILS_SUCCESS",
  USER_DETAILS_FAILURE: "USER_DETAILS_FAILURE",
  RESET_USER_DETAILS: "RESET_USER_DETAILS",
};
