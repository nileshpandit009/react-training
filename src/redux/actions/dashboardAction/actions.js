import { DASHBOARD_CONSTANTS } from "./actionTypes";

const {
  FETCH_POSTS,
  FETCH_POSTS_SUCCESS,
  FETCH_POSTS_ERROR,

  FETCH_POST_DETAILS,
  FETCH_POST_DETAILS_SUCCESS,
  FETCH_POST_DETAILS_ERROR,

  RESET_POSTS,
  RESET_POST_DETAILS,
} = DASHBOARD_CONSTANTS;

export const fetchPosts = (payload) => ({
  type: FETCH_POSTS,
  payload,
});

export const fetchPostsSuccess = (payload) => ({
  type: FETCH_POSTS_SUCCESS,
  payload,
});

export const fetchPostsError = (payload) => ({
  type: FETCH_POSTS_ERROR,
  payload,
});

export const fetchPostDetails = (payload) => ({
  type: FETCH_POST_DETAILS,
  payload,
});

export const fetchPostDetailsSuccess = (payload) => ({
  type: FETCH_POST_DETAILS_SUCCESS,
  payload,
});

export const fetchPostDetailsError = (payload) => ({
  type: FETCH_POST_DETAILS_ERROR,
  payload,
});

export const resetPosts = () => ({
  type: RESET_POSTS,
  payload: [],
});

export const resetPostDetails = () => ({
  type: RESET_POST_DETAILS,
  payload: {},
});
