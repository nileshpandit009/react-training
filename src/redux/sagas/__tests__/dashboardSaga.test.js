import {
  fetchPosts,
  fetchPostDetails,
  fetchPostsSaga,
  fetchPostDetailsSaga,
} from "redux/sagas/dashboardSaga";
import { put, takeEvery } from "@redux-saga/core/effects";
import {
  fetchPostDetailsSuccess,
  fetchPostsError,
  fetchPostsSuccess,
  fetchPostDetails as fetchPostDetailsAction,
  fetchPostDetailsError,
} from "redux/actions/dashboardAction/actions";
import { DASHBOARD_CONSTANTS } from "redux/actions/dashboardAction/actionTypes";

describe("tests dashboard saga for fetchPosts", () => {
  let fetchPostsGenerator;
  beforeEach(() => {
    fetchPostsGenerator = fetchPosts();
    const callDescriptor = fetchPostsGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it("must call fetchPostsSuccess action", () => {
    const expectedResponse = {
      data: [
        {
          id: 1,
          title: "abc",
          body: "abcabc",
        },
        {
          id: 2,
          title: "pqr",
          body: "pqrpqr",
        },
      ],
    };
    const putDescriptor = fetchPostsGenerator.next(expectedResponse).value;
    expect(putDescriptor).toEqual(
      put(fetchPostsSuccess({ posts: expectedResponse.data }))
    );

    expect(fetchPostsGenerator.next().done).toBeTruthy();
  });

  it("must call fetchPostsError action", () => {
    const errResponse = {
      status: 400,
      response: { data: "Some error message" },
    };
    const putDescriptor = fetchPostsGenerator.throw(errResponse).value;
    expect(putDescriptor).toEqual(
      put(fetchPostsError({ error: errResponse.response.data }))
    );

    expect(fetchPostsGenerator.next().done).toBeTruthy();
  });
});

describe("tests dashboard saga for fetchPostDetails", () => {
  let fetchPostDetailsGenerator;
  beforeEach(() => {
    fetchPostDetailsGenerator = fetchPostDetails(
      fetchPostDetailsAction({ id: 123 })
    );
    const callDescriptor = fetchPostDetailsGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it("must call fetchPostDetailsSuccess action", () => {
    const mockResponse = {
      data: {
        id: 1,
        title: "abc",
        body: "abcabc",
      },
    };

    const putDescriptor = fetchPostDetailsGenerator.next(mockResponse).value;
    expect(putDescriptor).toEqual(
      put(fetchPostDetailsSuccess({ postDetails: mockResponse.data }))
    );

    expect(fetchPostDetailsGenerator.next().done).toBeTruthy();
  });

  it("must call fetchPostDetailsError action", () => {
    const errResponse = {
      status: 400,
      response: { data: "Some error message" },
    };
    const putDescriptor = fetchPostDetailsGenerator.throw(errResponse).value;
    expect(putDescriptor).toEqual(
      put(fetchPostDetailsError({ error: errResponse.response.data }))
    );

    expect(fetchPostDetailsGenerator.next().done).toBeTruthy();
  });
});

describe("Test dashboard saga watchers", () => {
  let fetchPostsSagaItr;
  let fetchPostDetailsSagaItr;
  beforeAll(() => {
    fetchPostsSagaItr = fetchPostsSaga();
    fetchPostDetailsSagaItr = fetchPostDetailsSaga();
  });

  it("must start task to watch for fetchPosts action", () => {
    const takeEveryDescriptor = fetchPostsSagaItr.next().value;
    expect(takeEveryDescriptor).toEqual(
      takeEvery(DASHBOARD_CONSTANTS.FETCH_POSTS, fetchPosts)
    );

    expect(fetchPostsSagaItr.next().done).toBeTruthy();
  });

  it("must start task to watch for fetchPostDetails action", () => {
    const takeEveryDescriptor = fetchPostDetailsSagaItr.next().value;
    expect(takeEveryDescriptor).toEqual(
      takeEvery(DASHBOARD_CONSTANTS.FETCH_POST_DETAILS, fetchPostDetails)
    );

    expect(fetchPostDetailsSagaItr.next().done).toBeTruthy();
  });
});
