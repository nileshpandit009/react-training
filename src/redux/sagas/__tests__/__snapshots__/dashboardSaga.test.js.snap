// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests dashboard saga for fetchPostDetails must call fetchPostDetailsError action 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      123,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`tests dashboard saga for fetchPostDetails must call fetchPostDetailsSuccess action 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      123,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`tests dashboard saga for fetchPosts must call fetchPostsError action 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`tests dashboard saga for fetchPosts must call fetchPostsSuccess action 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
